packages/nextjs/app/builds/_components/AllBuilds.tsx

```diff
@@ -22,17 +24,51 @@ export function AllBuilds({ searchParams }: { searchParams: { category?: BuildCa
   const {
     data: builds,
     isLoading,
+    isFetching,
     refetch,
-  } = useQuery({
+    fetchNextPage,
+    hasNextPage,
+  } = useInfiniteQuery({
     queryKey: ["all-builds", debouncedFilter, categoryFilter, typeFilter],
-    queryFn: () =>
-      fetchBuilds({
+    queryFn: async ({ pageParam = 0 }) => {
+      const start = (pageParam as number) * FETCH_SIZE;
+      const response = await fetchBuilds({
         name: debouncedFilter,
         category: categoryFilter as BuildCategory,
         type: typeFilter as BuildType,
-      }),
+        start,
+        size: FETCH_SIZE,
+      });
+      return response;
+    },
+    initialPageParam: 0,
+    getNextPageParam: (lastGroup, groups) => {
+      if (lastGroup.data.length < FETCH_SIZE) {
+        return undefined;
+      }
+      return groups.length;
+    },
+    refetchOnWindowFocus: false,
+    placeholderData: keepPreviousData,
   });
 
+  const flatData = useMemo(() => builds?.pages?.flatMap(page => page.data) ?? [], [builds]);
+  const totalBuilds = builds?.pages?.[0]?.meta?.totalRowCount ?? 0;
+
+  useEffect(() => {
+    const onscroll = () => {
+      const scrolledTo = window.scrollY + window.innerHeight;
+      const isReachBottom = document.body.scrollHeight === scrolledTo;
+      if (isReachBottom && hasNextPage && !isFetching) {
+        fetchNextPage();
+      }
+    };
+    window.addEventListener("scroll", onscroll);
+    return () => {
+      window.removeEventListener("scroll", onscroll);
+    };
+  }, [fetchNextPage, hasNextPage, isFetching]);
```

Comment:
Maybe we can abstract the whole thing in separate hook file? 

maybe `useAllBuildsInfiniteQuery` ? And we can also flatData in that hook itself and just return data in well formated way and show it directly on the component?